# <center> Hello, World </center>

<img src=https://sdtimes.com/wp-content/uploads/2019/07/green-tree-python-1014229_640-490x326.jpg >

Did you know that Python was not named after this cute reptile?

*'When he began implementing Python, Guido van Rossum [the creator of the Python] was also reading the published scripts from “Monty Python's Flying Circus”, a BBC comedy series from the 1970s. Van Rossum thought he needed a name that was short, unique, and slightly mysterious, so he decided to call the language Python'.* [Python FAQ](https://docs.python.org/3/faq/general.html)

This notebook will help you brush up on your Python basics.

Some useful links, so you can recall the material, or find out what you did not know:

1. **Python documentation** https://docs.python.org/3/
2. **Python official and very long tutorial** https://docs.python.org/3/tutorial/ However, you will find all the information you need in our course there. I recommend to pay attention only to the chapters: 3. Informal Introduction to Python - 9. Classes.

Estimated reading time for tutorial's selected chapters: 3-5 hours. Solid Python programming skills - priceless. 

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


# Preliminaries. Fancy use of Jupyter Notebook

Here is a [cheat sheet](https://medium.com/edureka/jupyter-notebook-cheat-sheet-88f60d1aca7) on Jupyter Notebook. Take a look at the menu bar first. Make sure you know how to find the key operations in File: Open, Make a Copy, Save as, Rename, Download as.
On the bottom bar - check that you remember how to:
* **Save the current version of the notebook** (floppy disk icon). If you've been working in your notebook for too long, it's better to click on this icon and save a checkpoint (= an intermediate version of your file). Otherwise, if something goes wrong (connection is lost, your notebook hovers because can't execute your beautiful code anymore, or for any other reason when the universe creates new obstacles for you), your code will not be saved and will be lost. **Shortcut: Ctrl + S / Command + S**

* **Add a new cell** (plus icon). Don't code everything in one cell, organize and decompose it. **Shortcuts: Add a cell above: Ctrl + A / Command + A. Add a cell bellow: Ctrl + B / Command + B**

* Run a cell (this button is even named there for you). **Shortcut: Shift + Enter**

* Delete a cell. **Shortcut: D, D** - double click on D in a command mode.

* Interrupt the notebook (black square icon) and restart (circular arrow). In most cases, you use it if: 1) you forgot or did nor realize that your cell's output is too long and heavy to display; 2) you wrote something wrong, for instance, an infinite loop that never ends. **Shortcut to interupt the kernel: I, I** - double-click on the letter I. 

* Switch to markdown (code cell -> text cell) **M** in command mode or in the menu panel above. If you want to add titles or descriptions to your code, it's better to use markdown instead of the comments (#). 

! Before using the shortcuts, you need to click somewhere inside the notebook but outside the gray cells for coding/markdown. This is how you activate command mode. An alternative option is to enter the command mode (if you're currently in a gray code cell) - press Esc on your keyboard.

# Let's revise together

## Table of Contents
1. [Print](#print)
2. [Data types](#datatypes)
3. [Data structures](#datastructures)
4. [Indexing and slicing](#indexing)
5. [Data structures methods](#methods)

## Print <a name="print"></a>
print( ) is a function in Python3. It prints objects to the text stream file, separated by *sep* and followed by *end*. 

In [2]:
# output returns a string 
'Hello, world'

'Hello, world'

In [3]:
# output returns the result of print () function
print('Hello, world')

Hello, world


In [4]:
# compare these data types
type('Hello, world'), type(print('Hello, world'))

Hello, world


(str, NoneType)

This is because print( ) displays the output as you requested, but does not store it, and its type is NoneType.

Why are we using print( )?
This is useful if you want to see the output of the code above the last line:

In [5]:
# \n - new line
'If you are a nerd, \n'
print('Python is easy')

Python is easy


Only the last line is printed. What if we want both of them out? Now change the place for print ():

In [6]:
print('If you are a nerd,')
'Python is easy'

If you are a nerd,


'Python is easy'

In [7]:
# make sure you understand every character on that line
# we have 2 arguments (two string lines) in print()
print('And I think to myself', 'what a wonderful world', sep='\n', end='!')

And I think to myself
what a wonderful world!

*Sep* = separator; character separating arguments inside the print() function.

*End* - you can add characters at the end of your last argument in print( ).

[Show me more on print!](https://realpython.com/python-print/) *In fact, it contains too much information, so the first time you read this manual, please skip it.*

## Data types <a name="datatypes"></a>

The main types of numbers: **integer** (int), **float**. https://docs.python.org/3/tutorial/introduction.html#numbers

In [8]:
type(10)

int

In [9]:
type(10.0)

float

In [10]:
# float into integer
int(10.8)

10

Wait a minute, where is 0.8?

In [11]:
# mathematical rounding
round(10.8)

11

Converting a floating point number to an integer is not equal to mathematical rounding.

In [12]:
# the absolute value
abs(-100)

100

Arithmetical operations

In [13]:
2 + 2

4

In [14]:
print('Sum: 2 + 2 =', 2 + 2)

Sum: 2 + 2 = 4


In [15]:
print( 'Subtraction: 2 - 2 =', 2 - 2, 
      '\nDevision: 11 / 5 =', 11 / 5,
      '\nInteger devision: 11 // 5 =', 11 // 5,
      '\nMultiplication: 2 * 3 =', 2 * 3,
      '\nPower: 2 ** 3 =', 2 ** 3,
      '\n Remainder of division: 41 % 10 =', 41 % 10,
      sep=' ')

Subtraction: 2 - 2 = 0 
Devision: 11 / 5 = 2.2 
Integer devision: 11 // 5 = 2 
Multiplication: 2 * 3 = 6 
Power: 2 ** 3 = 8 
 Remainder of division: 41 % 10 = 1


For text data we use **strings**. https://docs.python.org/3/tutorial/introduction.html#strings

In [16]:
profession = 'Fashion'

print(profession, 'is my profession.')
type(profession)

Fashion is my profession.


str

**Bool**

In [17]:
type(True)

bool

In [18]:
True or False

True

In [19]:
# True = 1, False = 0
1 or 0

1

In [20]:
True and False

False

In [21]:
# = means to store the value (5) with a pointer ('number') variable in memory
number = 5

# == means 'equal to' and returns True or False
number == 5

True

!= means 'not equal to'

In [22]:
number != 5

False

In [23]:
number <= 5

True

## Data structures <a name="datastructures"></a>

While reading, pay attention to the types of brackets.

<img src="./Rick_and_Morty.jpg" width=400 height=300 align='left'>

**Lists** and arrays are the same thing in Python. https://docs.python.org/3/tutorial/introduction.html#lists and https://docs.python.org/3/tutorial/datastructures.html#more-on-lists

In [24]:
supermarket_queue = ['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']
type(supermarket_queue)

list

In [25]:
# Python keeps the elements of list in the same order
supermarket_queue

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

**Dictionaries** consists of pairs: { key : value, key : value, ... }. 

Dictionary values can be of any type, but the keys must be of an immutable data type such as strings, numbers, or tuples.

https://docs.python.org/3/tutorial/datastructures.html#dictionaries

In [26]:
family_age_info = {'You': 20, 'Summer' : 17, 'Jerry' : 35, 'Rick' : 70, 'Morty' : 14, 
                   'Beth' : 34, 'You' : 'forever young'}
type(family_age_info)

dict

In [27]:
family_age_info

{'You': 'forever young',
 'Summer': 17,
 'Jerry': 35,
 'Rick': 70,
 'Morty': 14,
 'Beth': 34}

The keys in dictionaries must be unique. If you have assigned several values to one key in a row, the dictionary will keep the last value. This is why {'You' : 20} is overwritten to {'You' : 'forever young'}.

**Tuple**

https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences

In [28]:
rick_nicknames = ('Grandpa Rick', 'Dad', 'Rick', 'Albert Ein-douche', 
                  'Pickle Rick', 'Smartest Man in the Universe')
type(rick_nicknames)

tuple

In [29]:
rick_nicknames

('Grandpa Rick',
 'Dad',
 'Rick',
 'Albert Ein-douche',
 'Pickle Rick',
 'Smartest Man in the Universe')

Let's change one of the nicknames:

In [30]:
rick_nicknames[0] = 'Other nickname'

TypeError: 'tuple' object does not support item assignment

The **TypeError** was raised because, unlike lists, tuples are immutable. This is why we cannot change the assigned items.

**Set** 

We usually use sets instead of lists and tuples when we need to check if an element is in a set of unique elements or not. Sets as a data structure are faster than others.
https://docs.python.org/3/tutorial/datastructures.html#sets

In [32]:
mood_diary_during_month = {'gloomy', 'happy', 'sad', 'annoyed', 'scared', 
                           'happy', 'glad', 'sleepy', 'gloomy', 'annoyed', 'romantic', 
                           'depressed', 'sleepy', 'scared', 'annoyed', 'happy', 'romantic',
                           'scared', 'sleepy', 'annoyed', 'scared', 'happy', 'glad',
                          'gloomy', 'annoyed', 'romantic', 'happy', 'gloomy', 'gloomy',
                          'happy', 'sad', 'annoyed', 'depressed', 'sleepy', 'scared',
                          'annoyed', 'scared', 'gloomy', 'annoyed', 'romantic'}

In [33]:
# only unique elements are kept
mood_diary_during_month

{'annoyed',
 'depressed',
 'glad',
 'gloomy',
 'happy',
 'romantic',
 'sad',
 'scared',
 'sleepy'}

In [34]:
# check if "annoying" is one of the elements in the set
'annoyed' in mood_diary_during_month

True

Summing up the properties. **Ordered** means that Python stores elements in the same order you specified when you created it. **Mutable** means that we can change the elements, for example, to replace one of the elements with a new value. **Constructor** demonstrates how to create, represent and store each data type. Note the difference in brackets.

<img src="https://miro.medium.com/max/1400/1*Det-kkoSw9T4IZ4XrypVNQ.png" style="width:700px;height:300px"/>

## Indexing and slicing <a name="indexing"></a>

In [35]:
supermarket_queue

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

In [36]:
# list length
len(supermarket_queue)

5

Each item in the list has its own index. 0 - 'Summer', 1 - 'Jerry', etc. You can return an item from a list by its index using square brackets [ ].

<img src="https://cdn.programiz.com/sites/tutorial2program/files/python-list-index.png" width="500" height="500" align="left">

In [37]:
# who's first in line?
# indexing in Python starts at 0 instead of 1

supermarket_queue[0]

'Summer'

In [38]:
# who is the last one?

supermarket_queue[-1]

'Morty'

or

In [39]:
# the index of the last element is (length - 1) because indices start at 0 instead of 1

supermarket_queue[len(supermarket_queue) - 1]

'Morty'

Explanation: **[start index : end index : step]** [more details](https://stackoverflow.com/questions/509211/understanding-slice-notation)

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

    0      1    2     3    4

In [40]:
# all elements starting from the second element
supermarket_queue[1:]

['Jerry', 'Beth', 'Rick', 'Morty']

In [41]:
# all elements before the last element
supermarket_queue[:-1]

['Summer', 'Jerry', 'Beth', 'Rick']

In [42]:
# all elements with one step skip (keep elements at even indices only)
supermarket_queue[::2]

['Summer', 'Beth', 'Morty']

In [43]:
# all elements in reverse order
supermarket_queue[::-1]

['Morty', 'Rick', 'Beth', 'Jerry', 'Summer']

In [44]:
supermarket_queue

['Summer', 'Jerry', 'Beth', 'Rick', 'Morty']

In [45]:
# alternative way to reverse list in place
supermarket_queue.reverse()
supermarket_queue

['Morty', 'Rick', 'Beth', 'Jerry', 'Summer']

In [46]:
# nested list indexing
supermarket_queue.append(['Bad Snuffles', 'Good Snuffles', 'Alternative Snuffles'])
supermarket_queue

['Morty',
 'Rick',
 'Beth',
 'Jerry',
 'Summer',
 ['Bad Snuffles', 'Good Snuffles', 'Alternative Snuffles']]

In [47]:
# the sixth item in the list to go to the nested list, get the second item in the nested list
supermarket_queue[5][1]

'Good Snuffles'

## Data structures methods (lists, strings, tuples, dictinonaries, sets) <a name="methods"></a>

### Lists

In [48]:
# Grades for HA1, HA2, Project
your_potential_grades = [10, 9.5, 10]

# Look up your second grade (HA2)
your_potential_grades[1] 

9.5

In [49]:
# run the cell above
# in this cell try to type 'yo' and click on TAB on your keybord


Magic. Jupyter Notepad helps you with autocomplete for long variable names.

Let's sort the grade **in ascending order** in place (it means, that this list will be stored as sorted). 

In [50]:
# sorting 

your_potential_grades.sort()
your_potential_grades

[9.5, 10, 10]

In [51]:
# click inside the brackets of sort and push Shift + Tab
your_potential_grades.sort()

This is how you call the documentation for each function in Python. So, you can see, there are two arguments: key and reverse. Let's use reverse and sort our list **in descending order**.

In [52]:
your_potential_grades.sort(reverse=True)
your_potential_grades

[10, 10, 9.5]

In [53]:
# alternative with creteing a copy of the list

your_potential_grades = [10, 9.5, 10]

sorted(your_potential_grades)

[9.5, 10, 10]

In [54]:
your_potential_grades

[10, 9.5, 10]

Why didn't Python keep your_potential_grades sorted and return the original order of the elements? This is because sorted () creates a copy of your list and then sorts it.

In [55]:
sorted_grades = sorted(your_potential_grades)
sorted_grades

[9.5, 10, 10]

Your can change the grade you don't like, because lists are mutable. 

In [56]:
# replace
your_potential_grades[1] = 10
your_potential_grades[1]

10

Imagine, you liked this course so much, that you even completed our bonus task and got 2.5 additional points. 

In [57]:
# add element 
your_potential_grades.append(2.5)

your_potential_grades

[10, 10, 10, 2.5]

In [58]:
# delete element
your_potential_grades.remove(2.5)

your_potential_grades

[10, 10, 10]

In [59]:
# pop element
your_potential_grades.pop()

10

In [60]:
your_potential_grades

[10, 10]

pop() returns the last item in the list and removes it, while the remove() method removes the item you specified in brackets (with whatever index you want) and does not return that value on output.

In [61]:
# you can save the result of pop() into a variable
last_grade_in_list = your_potential_grades.pop()
last_grade_in_list

10

In [62]:
# insert an element at a specific position
# 0 - index, 10 - value
your_potential_grades.insert(0, 9)
your_potential_grades

[9, 10]

In [63]:
your_potential_grades.insert(1, 10)
your_potential_grades

[9, 10, 10]

In [64]:
# let's count how many times the value 10 is in the list
your_potential_grades.count(10)

2

### List comprehensions

In [65]:
some_list = [0, 1, 2, 3, 4, 5, 6]

new_list = []
for i in some_list:
    new_list.append(i ** 2) # square 
    
new_list

[0, 1, 4, 9, 16, 25, 36]

Instead of loop with **for**, we can code this operation in one line using list comprehensions:

In [66]:
another_new_list = [i ** 2 for i in some_list]
another_new_list

[0, 1, 4, 9, 16, 25, 36]

The square brackets around the expression should indicate that we are creating a list. The expression inside the brackets should be read literally: a list of squared elements for (**for**) elements i from (**in**) some_list.

In [67]:
# filtering using list comprehensions
[x for x in another_new_list if x % 2 == 0]

[0, 4, 16, 36]

% operator returns the remainder of the division. So, ```if x % 2 == 0``` means that if the number x is divisible by 2 and has no remainder (remainder = 0 is True), then we store it in the list, otherwise, delete it. As a result, we get even numbers.

### Strings

The key operations: concatenation, split, join and extending string with its duplicates by *N* times. 

In [68]:
greeting = 'welcome to the course on Data Analysis!'

whom = 'Dear students, '

# concatenation
print(whom + greeting)

Dear students, welcome to the course on Data Analysis!


In [69]:
# repeat by mulyiplicator
print('You will need to work', 'very ' * 5, 'hard...')

You will need to work very very very very very  hard...


In [70]:
# string starts with
print(greeting.startswith('Chilling'))

False


In [71]:
# string ends with
print(greeting.endswith('sis!'))
print('Recheck it'.endswith('t'))

True
True


In [72]:
# split
text_line = 'You need_to split this_sentence into four parts_by underscore sign.'
print(text_line.split('_'))

['You need', 'to split this', 'sentence into four parts', 'by underscore sign.']


In [73]:
# join
text_line = 'You need to join this sentence by asterisks instead of white spaces.'
'*'.join(text_line.split(' '))

'You*need*to*join*this*sentence*by*asterisks*instead*of*white*spaces.'

The find( ) method returns the index of the first character of the first occurrence of the substring, if contained in the string. If there is no such substring, it returns -1.

In [74]:
# find

prologue = '''The morning had dawned clear and cold, with a crispness that hinted at
the end of summer. They set forth at daybreak to see a man beheaded, twenty in all, and
Bran rode among them, nervous with excitement. This was the first time he had been deemed
old enough to go with his lord father and his brothers to see the king’s justice done. It was
the ninth year of summer, and the seventh of Bran’s life'''

prologue.find('Bran')

159

The name 'Bran' was found and 159 is the index of the first letter of this word in a string 'prologue'. 

In [75]:
print(prologue[prologue.find('Bran'):]) # prints the whole string starting from 'Bran'

Bran rode among them, nervous with excitement. This was the first time he had been deemed
old enough to go with his lord father and his brothers to see the king’s justice done. It was
the ninth year of summer, and the seventh of Bran’s life


In [76]:
print(prologue[prologue.find('Bran'):prologue.find(' rode')]) #slicing

Bran


Let's make a simple version of sentiment analysis from scratch. If the review contains the word 'bad', then we will write that this is a negative review.

In [77]:
feedback = 'This place was bad enough.'

print(feedback)
if feedback.find('bad') == -1: # if 'bad' not found
    print('Feedback is not negative')
else:
    print('Feedback is negative')

This place was bad enough.
Feedback is negative


In [78]:
# replace
statement = '''I take you, Monica, to be my wife.\
I promise to be true to you in good times and in bad, in sickness and in health.\
I will love you and honour you all the days of my life.'''

statement.replace('Monica', 'Rachel')

'I take you, Rachel, to be my wife.I promise to be true to you in good times and in bad, in sickness and in health.I will love you and honour you all the days of my life.'

### Tuples

https://www.w3schools.com/python/python_tuples.asp

In [79]:
info = ('Sherlock Holmes', 'Baker Street 221b', 'Deduction', 'Elementary', 'Dr. Watson')
type(info)

tuple

In [80]:
# access item
info[3]

'Elementary'

In [81]:
info[-1]

'Dr. Watson'

In [82]:
# slicing
info[2:5]

('Deduction', 'Elementary', 'Dr. Watson')

In [83]:
info[2:100000000] # the upper bound of index in slicing is unlimited(the same for lists and strings)

('Deduction', 'Elementary', 'Dr. Watson')

In [84]:
# replacement doesn't work because tuples are immutable
info[2] = 'Induction'
info

TypeError: 'tuple' object does not support item assignment

However, you can convert your tuple into list, make change and convert it back. 

In [85]:
info = list(info)
info

['Sherlock Holmes',
 'Baker Street 221b',
 'Deduction',
 'Elementary',
 'Dr. Watson']

In [86]:
info[2] = 'Induction'
info = tuple(info)
info

('Sherlock Holmes',
 'Baker Street 221b',
 'Induction',
 'Elementary',
 'Dr. Watson')

**Once a tuple is created, you cannot add and remove items in it**. Tuples are unchangeable, but you can use the trick with list and convert it back.

### Dictionaries

A dictionary is an unordered structure, it consists of keys and values. We cannot refer to the object by index, but we can refer to the key.

In [87]:
car = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}

We cannot refer to the object by index, but we can refer to the key:

In [88]:
car[0]

KeyError: 0

In [89]:
car['model']

'Mustang'

To find out what keys or values are in the dictionary, you can use the corresponding methods.

In [90]:
car.keys()

dict_keys(['brand', 'model', 'year'])

In [91]:
car.values()

dict_values(['Ford', 'Mustang', 1964])

In [92]:
car.items()

dict_items([('brand', 'Ford'), ('model', 'Mustang'), ('year', 1964)])

In [93]:
for key, value in car.items():
    print('key:', key, '- value:', value)

key: brand - value: Ford
key: model - value: Mustang
key: year - value: 1964


In [94]:
for key in car:
    print(key)

brand
model
year


Adding key-value pairs to a dictionary is very simple, like replacing a value by index in lists. The deletion is done using the ```del``` function. We can also check the presence of the key in the dictionary.

In [95]:
car['colour'] = 'red' # new key and value
car['layout'] = 'front-engine'
car['year'] = '1967'  # overwrite the value in existing key
print(car)

{'brand': 'Ford', 'model': 'Mustang', 'year': '1967', 'colour': 'red', 'layout': 'front-engine'}


In [96]:
del car['layout']
print(car)

{'brand': 'Ford', 'model': 'Mustang', 'year': '1967', 'colour': 'red'}


In [97]:
'brand' in car # is 'brand' among keys?

True

### Sets

<img src="https://www.learnbyexample.org/wp-content/uploads/python/Python-Set-Operatioons.png" height=300 width=300>

Some names are only used for girls, some are only used for both, but some are used for both (unisex).

In [98]:
boys_names = {'James', 'Alexis', 'Oliver', 'Gale', 'Taylor', 'Benjamin'}
girls_names = {'Maria', 'Alexis', 'Ann', 'Gale', 'Clara', 'Emma', 'Sophia', 'Valeria', 'Polina'}

print(boys_names | girls_names)  # union of sets -- we join all unique names from two sets together
print(boys_names & girls_names)  # intersection -- keep the elements that are in both sets
print(boys_names - girls_names)  # difference -- elements in the first set, but not in the second
print(boys_names ^ girls_names)  # symmetric difference -- elements from A | B, but not from A & B

{'Maria', 'Clara', 'Sophia', 'Polina', 'Alexis', 'Benjamin', 'James', 'Taylor', 'Emma', 'Ann', 'Gale', 'Oliver', 'Valeria'}
{'Alexis', 'Gale'}
{'Taylor', 'Oliver', 'Benjamin', 'James'}
{'Maria', 'Clara', 'Sophia', 'Polina', 'James', 'Emma', 'Benjamin', 'Taylor', 'Ann', 'Oliver', 'Valeria'}


Instead of signs as operators you can use methods:

In [99]:
print(boys_names.union(girls_names))
print(boys_names.intersection(girls_names))
print(boys_names.difference(girls_names))
print(boys_names.symmetric_difference(girls_names))

{'Maria', 'Clara', 'Sophia', 'Polina', 'Alexis', 'Benjamin', 'James', 'Taylor', 'Emma', 'Ann', 'Gale', 'Oliver', 'Valeria'}
{'Alexis', 'Gale'}
{'Taylor', 'Oliver', 'Benjamin', 'James'}
{'Maria', 'Clara', 'Sophia', 'Polina', 'James', 'Emma', 'Benjamin', 'Taylor', 'Ann', 'Oliver', 'Valeria'}
